controlplane/internal/controllers/scale.go

```diff
@@ -151,20 +151,6 @@ func (r *RKE2ControlPlaneReconciler) scaleDownControlPlane(
 		return ctrl.Result{}, errors.New("failed to pick control plane Machine to delete")
 	}
 
-	// If etcd leadership is on machine that is about to be deleted, move it to the newest member available.
```

Comment:
Potentially this change can also check that more than 1 CP machine is not currently tagged for deletion, and perform immediate etcd leadership transfer only on scaling down to 1 CP. Then the regular memebership removal can be done in a [periodic](https://github.com/rancher/cluster-api-provider-rke2/blob/v0.6.0/controlplane/internal/controllers/rke2controlplane_controller.go#L511-L515). But I can’t tell if the leader removed on >2 replicas can be transferred at the moment of scale down to 1.

We should be testing this scenario in e2e, unless there were changes.

Reply:
You're right, it would probably be preferable to keep the immediate etcd leadership transfer part, as it'll prevent transcient Leader Failures documented [here](https://github.com/improbable-eng/etcd-cluster-operator/blob/master/docs/operations.md#scale-down-operations).

There could indeed be some regression on https://github.com/rancher/cluster-api-provider-rke2/issues/263, but on the other hand, I tend to believe that this issue is way less critical than having rolling updates of control planes blocked by node drain. 

To that extent, current e2e test are not representative of typical deployments, because as nodeDrainTimeout is [set to 30s](https://github.com/rancher/cluster-api-provider-rke2/blob/main/test/e2e/data/infrastructure/cluster-template-docker.yaml#L119), the described issue has no chance to appear, as nodes will be deleted quickly by capi controller. But in that case we give only few changes to statefull applications to handover in a clean way, and podDisruptionBudgets won't be honored.

That being said, from what I understand, etcd will become read-only when control place is scaled down from 2 to 1 as it will loose its chorum. But in that case, API server should be still available, couldn't we just figure ou a way to ensure that rke2-controller will succes to remove the member corresponding to the deleted machine anyway in that case?

Reply:
While searching for pre-terminate hook in cluster-api code, I found that PR that landed a few days ago: https://github.com/kubernetes-sigs/cluster-api/pull/11137.
It seems to be implementing something similar to the second proposition above, it could be a nice example to follow!

Reply:
Closing that PR in flavor of https://github.com/rancher/cluster-api-provider-rke2/pull/435 as it is adressing the issue in a better way (following https://github.com/kubernetes-sigs/cluster-api/pull/11137 as discussed above)